### PR TITLE
(17.5) Copy all diagnostic info from derived classes when changing severity of a diagnostic

### PR DESCRIPTION
**Pull request by RikkiGibson**
Backport of #66038 to 17.5

@jcouv @jaredpar @Cosifne for visibility.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs [hunk 1]
```diff
@@ -2770,5 +2770,41 @@ public class C
...
+}", options: options);
+
+            var diagnostics = comp.GetDiagnostics();
+            diagnostics.Verify(
+                // (8,14): warning CS8618: Non-nullable field 'f' must contain a non-null value when exiting constructor. Consider declaring the field as nullable.
+                //     public B f;
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "f").WithArguments("field", "f").WithLocation(8, 14).WithWarningAsError(warnAsError));
+
+            var diagnostic = Assert.Single(diagnostics);
+            Assert.Single(diagnostic.AdditionalLocations);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs:2806]: >Assert.Single(diagnostic.AdditionalLocations) [](http://example.com/codeflow?start=12&length=45)

If the original diagnostic included symbols, consider adding:
```
Assert.Equal(new[] { "...", ... }, diagnostic.Symbols.Select(s => s.ToTestDisplayString()));
```

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs:2806]: The diagnostic doesn't include symbols so I think I won't exercise this part for now.

**Review by cston — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Comment by RikkiGibson** [comment]
Ugh. My brain wasn't working properly. Have to do QB paperwork. Do you need me to revert this or is it fine to just proceed with the QB paperwork for now?

**Pull request merged**